### PR TITLE
Add DNS Provider for DreamHost

### LINKED ISSUE #667: Add DNS Challenge support for domains hosted by DreamHost
[DreamHost](https://www.dreamhost.com/domains/) offers DNS Management for domains registered through them, and offers an [API](https://help.dreamhost.com/hc/en-us/articles/217560167-Application-programming-interface-overview).

### PR DESCRIPTION
**Pull request by cjs**
Fixes #667.

This adds DNS Provider support for [Dreamhost](https://www.dreamhost.com/domains/) through their [API](https://help.dreamhost.com/hc/en-us/articles/217555707-DNS-API-commands).

In addition to the tests included in the PR, I manually live tested. The only issue is DreamHost has very slow propogation. In my test case it took around 50 minutes for the record to propagate.

### CODE REVIEW

File: providers/dns/dreamhost/dreamhost.go [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+	BaseURL            string
+	APIKey             string
+	PropagationTimeout time.Duration
+	PollingInterval    time.Duration
+	HTTPClient         *http.Client
+}
+
+// NewDefaultConfig returns a default configuration for the DNSProvider
+func NewDefaultConfig() *Config {
+	client := acme.HTTPClient
```
Comment (ldez) [inline providers/dns/dreamhost/dreamhost.go]: could you rewrite like that:
```go
func NewDefaultConfig() *Config {
	return &Config{
		BaseURL:            defaultBaseURL,
		PropagationTimeout: env.GetOrDefaultSecond("DREAMHOST_PROPAGATION_TIMEOUT", acme.DefaultPropagationTimeout),
		PollingInterval:    env.GetOrDefaultSecond("DREAMHOST_POLLING_INTERVAL", acme.DefaultPollingInterval),
		HTTPClient: &http.Client{
			Timeout: env.GetOrDefaultSecond("DREAMHOST_HTTP_TIMEOUT", 30*time.Second),
		},
	}
}
```

File: providers/dns/dreamhost/dreamhost.go [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+
+// Timeout returns the timeout and interval to use when checking for DNS propagation.
+// Adjusting here to cope with spikes in propagation times.
+func (d *DNSProvider) Timeout() (timeout, interval time.Duration) {
+	return d.config.PropagationTimeout, d.config.PollingInterval
+}
+
+// addTxtRecord adds the TXT record to the domain
+func (d *DNSProvider) updateTxtRecord(action, token, domain, txt string) error {
+	var cmd string
```
Comment (ldez) [inline providers/dns/dreamhost/dreamhost.go]: You can create some constants at the top of the file:
```go
const (
	cmdAddRecord    = "dns-add_record"
	cmdRemoveRecord = "dns-remove_record"
)
```

and also remove the switch-case.

File: providers/dns/dreamhost/dreamhost.go [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+	case "add":
+		cmd = "dns-add_record"
+	case "remove":
+		cmd = "dns-remove_record"
+	default:
+		return fmt.Errorf("dreamhost: updateTxtRecord called with invalid action: %s", action)
+	}
+
+	comment := url.QueryEscape("Managed By lego")
+	u := fmt.Sprintf("%s/?key=%s&cmd=%s&format=json&record=%s&type=TXT&value=%s&comment=%s", d.config.BaseURL, token, cmd, domain, txt, comment)
```
Comment (ldez) [inline providers/dns/dreamhost/dreamhost.go]: could you rewrite this part like that:

```go
u, err := url.Parse(d.config.BaseURL)
if err != nil {
	return err
}

query := u.Query()
query.Set("key", token)
query.Set("cmd", action)
query.Set("format", "json")
query.Set("record", domain)
query.Set("type", "TXT")
query.Set("value", txt)
query.Set("comment", url.QueryEscape("Managed By lego"))
u.RawQuery = query.Encode()
```

File: providers/dns/dreamhost/dreamhost.go [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+
+	resp, err := acme.HTTPClient.Get(u)
```
Comment (ldez) [inline providers/dns/dreamhost/dreamhost.go]:
```go
resp, err := d.config.HTTPClient.Get(u)
```

File: providers/dns/dreamhost/dreamhost.go [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+	if resp.StatusCode != 200 {
+		return fmt.Errorf("request failed with HTTP status code %d", resp.StatusCode)
+	}
+
+	var response responseStruct
+	err = json.NewDecoder(resp.Body).Decode(&response)
+	// return fmt.Errorf(fmt.Sprint(response.Result))
+	if err != nil {
+		return fmt.Errorf("unable to decode API server response")
+	} else if response.Result == "error" {
```
Comment (ldez) [inline providers/dns/dreamhost/dreamhost.go]: you can replace `else if` by a simple `if`

File: providers/dns/dreamhost/dreamhost.go [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != 200 {
+		return fmt.Errorf("request failed with HTTP status code %d", resp.StatusCode)
+	}
+
+	var response responseStruct
+	err = json.NewDecoder(resp.Body).Decode(&response)
+	// return fmt.Errorf(fmt.Sprint(response.Result))
```
Comment (ldez) [inline providers/dns/dreamhost/dreamhost.go]: could you delete that line.

File: providers/dns/dreamhost/dreamhost_test.go [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+var fakeDreamHostAPIKey = "asdf1234"
+var fakeDreamHostChallengeToken = "foobar"
+var fakeDreamHostKeyAuth = "w6uP8Tcg6K2QR905Rms8iXTlksL6OD1KOWBxTK7wxPI"
+
+func TestDreamHostPresent(t *testing.T) {
+	var requestReceived bool
```
Comment (ldez) [inline providers/dns/dreamhost/dreamhost_test.go]: could you remove that variable.

File: providers/dns/dreamhost/dreamhost_test.go [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+	config.BaseURL = mock.URL
+
+	provider, err := NewDNSProviderConfig(config)
+	require.NoError(t, err)
+	require.NotNil(t, provider)
+
+	err = provider.Present("example.com", "", fakeDreamHostChallengeToken)
+	require.NoError(t, err, "fail to create TXT record")
+
+	assert.True(t, requestReceived, "Expected request to be received by mock backend, but it wasn't")
```
Comment (ldez) [inline providers/dns/dreamhost/dreamhost_test.go]: could you remove that line.

File: providers/dns/dreamhost/dreamhost_test.go [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+}
+
+func TestDreamHostPresentFailed(t *testing.T) {
+	var requestReceived bool
```
Comment (ldez) [inline providers/dns/dreamhost/dreamhost_test.go]: could you remove that variable.

File: providers/dns/dreamhost/dreamhost_test.go [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+	require.NotNil(t, provider)
+
+	err = provider.Present("example.com", "", fakeDreamHostChallengeToken)
+	require.Error(t, err, "API error not detected")
+
+	assert.True(t, requestReceived, "Expected request to be received by mock backend, but it wasn't")
+}
+
+func TestDreamHostCleanup(t *testing.T) {
+	var requestReceived bool
```
Comment (ldez) [inline providers/dns/dreamhost/dreamhost_test.go]: could you remove that variable.

File: providers/dns/dreamhost/dreamhost_test.go [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+	config.BaseURL = mock.URL
+
+	provider, err := NewDNSProviderConfig(config)
+	require.NoError(t, err)
+	require.NotNil(t, provider)
+
+	err = provider.Present("example.com", "", fakeDreamHostChallengeToken)
+	require.Error(t, err, "API error not detected")
+
+	assert.True(t, requestReceived, "Expected request to be received by mock backend, but it wasn't")
```
Comment (ldez) [inline providers/dns/dreamhost/dreamhost_test.go]: could you remove that line.

File: providers/dns/dreamhost/dreamhost_test.go [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+	config.BaseURL = mock.URL
+
+	provider, err := NewDNSProviderConfig(config)
+	require.NoError(t, err)
+	require.NotNil(t, provider)
+
+	err = provider.CleanUp("example.com", "", fakeDreamHostChallengeToken)
+	require.NoError(t, err, "failed to remove TXT record")
+
+	assert.True(t, requestReceived, "Expected request to be received by mock backend, but it wasn't")
```
Comment (ldez) [inline providers/dns/dreamhost/dreamhost_test.go]: could you remove that line.

File: providers/dns/dreamhost/dreamhost_test.go [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+
+		q := r.URL.Query()
+
+		assert.Equal(t, q.Get("key"), fakeDreamHostAPIKey, "key mismatch")
+		assert.Equal(t, q.Get("cmd"), "dns-add_record", "cmd mismatch")
+		assert.Equal(t, q.Get("record"), "_acme-challenge.example.com")
+		assert.Equal(t, q.Get("value"), fakeDreamHostKeyAuth, "value mismatch")
+
+		w.WriteHeader(http.StatusOK)
+		fmt.Fprintf(w, `{"data":"record_added","result":"success"}`)
```
Comment (ldez) [inline providers/dns/dreamhost/dreamhost_test.go]:
```go
_, err := fmt.Fprintf(w, `{"data":"record_added","result":"success"}`)
if err != nil {
	http.Error(w, err.Error(), http.StatusInternalServerError)
}
```

File: providers/dns/dreamhost/dreamhost_test.go [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+func TestDreamHostPresentFailed(t *testing.T) {
+	var requestReceived bool
+
+	mock := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		requestReceived = true
+
+		assert.Equal(t, http.MethodGet, r.Method, "method")
+
+		w.WriteHeader(http.StatusOK)
+		fmt.Fprintf(w, `{"data":"record_already_exists_remove_first","result":"error"}`)
```
Comment (ldez) [inline providers/dns/dreamhost/dreamhost_test.go]:
```go
_, err := fmt.Fprintf(w, `{"data":"record_already_exists_remove_first","result":"error"}`)
if err != nil {
	http.Error(w, err.Error(), http.StatusInternalServerError)
}
```

File: providers/dns/dreamhost/dreamhost_test.go [hunk 15]
```diff
@@ -0,0 +1,110 @@
...
+
+		q := r.URL.Query()
+
+		assert.Equal(t, q.Get("key"), fakeDreamHostAPIKey, "key mismatch")
+		assert.Equal(t, q.Get("cmd"), "dns-remove_record", "cmd mismatch")
+		assert.Equal(t, q.Get("record"), "_acme-challenge.example.com")
+		assert.Equal(t, q.Get("value"), fakeDreamHostKeyAuth, "value mismatch")
+
+		w.WriteHeader(http.StatusOK)
+		fmt.Fprintf(w, `{"data":"record_removed","result":"success"}`)
```
Comment (ldez) [inline providers/dns/dreamhost/dreamhost_test.go]:
```go
_, err := fmt.Fprintf(w, `{"data":"record_removed","result":"success"}`)
if err != nil {
	http.Error(w, err.Error(), http.StatusInternalServerError)
}
```

File: providers/dns/dreamhost/dreamhost_test.go [hunk 16]
```diff
@@ -0,0 +1,110 @@
...
+		assert.Equal(t, http.MethodGet, r.Method, "method")
+
+		q := r.URL.Query()
+
+		assert.Equal(t, q.Get("key"), fakeDreamHostAPIKey, "key mismatch")
+		assert.Equal(t, q.Get("cmd"), "dns-remove_record", "cmd mismatch")
+		assert.Equal(t, q.Get("record"), "_acme-challenge.example.com")
+		assert.Equal(t, q.Get("value"), fakeDreamHostKeyAuth, "value mismatch")
+
+		w.WriteHeader(http.StatusOK)
```
Comment (ldez) [inline providers/dns/dreamhost/dreamhost_test.go]: could you remove that line.

File: providers/dns/dreamhost/dreamhost_test.go [hunk 17]
```diff
@@ -0,0 +1,110 @@
...
+
+func TestDreamHostPresentFailed(t *testing.T) {
+	var requestReceived bool
+
+	mock := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		requestReceived = true
+
+		assert.Equal(t, http.MethodGet, r.Method, "method")
+
+		w.WriteHeader(http.StatusOK)
```
Comment (ldez) [inline providers/dns/dreamhost/dreamhost_test.go]: could you remove that line.

File: providers/dns/dreamhost/dreamhost_test.go [hunk 18]
```diff
@@ -0,0 +1,110 @@
...
+		assert.Equal(t, http.MethodGet, r.Method, "method")
+
+		q := r.URL.Query()
+
+		assert.Equal(t, q.Get("key"), fakeDreamHostAPIKey, "key mismatch")
+		assert.Equal(t, q.Get("cmd"), "dns-add_record", "cmd mismatch")
+		assert.Equal(t, q.Get("record"), "_acme-challenge.example.com")
+		assert.Equal(t, q.Get("value"), fakeDreamHostKeyAuth, "value mismatch")
+
+		w.WriteHeader(http.StatusOK)
```
Comment (ldez) [inline providers/dns/dreamhost/dreamhost_test.go]: could you remove that line.

File: providers/dns/dreamhost/client.go [hunk 19]
```diff
@@ -0,0 +1,10 @@
+package dreamhost
+
+const defaultBaseURL = "https://api.dreamhost.com"
+
+// types for JSON responses
+
+type responseStruct struct {
```
Comment (ldez) [inline providers/dns/dreamhost/client.go]: could you rename `responseStruct` to `apiResponse`.

**Comment by cjs** [comment]
@ldez Thanks for detailed review, it helps very much. It might be fairly obvious that I wrote this based on the structure of some existing providers, and I used the `digitalocean` provider as a reference for the tests. 

Do you have a preference if the `consts` and `apiResponse` `struct` are defined in the `dreamhost.go` file or in the package `client.go` file?

**Comment by cjs** [comment]
```
$ env DREAMHOST_PROPAGATION_TIMEOUT=14400 DREAMHOST_POLLING_INTERVAL=60 \
./lego -m REDACTED -a -x http-01 -x tls-alpn-01 --dns dreamhost \
-s https://acme-staging-v02.api.letsencrypt.org/directory --dns-resolvers 8.8.8.8 \
-d readyfireaim.org -d '*.readyfireaim.org' run
2018/10/09 12:26:52 [INFO] [readyfireaim.org, *.readyfireaim.org] acme: Obtaining bundled SAN certificate
2018/10/09 12:26:53 [INFO] [*.readyfireaim.org] AuthURL: https://acme-staging-v02.api.letsencrypt.org/acme/authz/ZoLlfQXdu8uOVGOA59_QiyrKXaCtfbouvQHw3NzIr7Y
2018/10/09 12:26:53 [INFO] [readyfireaim.org] AuthURL: https://acme-staging-v02.api.letsencrypt.org/acme/authz/11svaJCTvsgWrQxBXGo1Ry5zFfzJJ6D0monkTg_wOKU
2018/10/09 12:26:53 [INFO] [readyfireaim.org] acme: Authorization already valid; skipping challenge
2018/10/09 12:26:53 [INFO] [readyfireaim.org] acme: Preparing to solve DNS-01
2018/10/09 12:26:54 [INFO] dreamhost: record_added
2018/10/09 12:26:54 [INFO] [readyfireaim.org] acme: Trying to solve DNS-01
2018/10/09 12:26:54 [INFO] [readyfireaim.org] Checking DNS record propagation using [8.8.8.8:53]
2018/10/09 12:26:54 [INFO] Wait [timeout: 4h0m0s, interval: 1m0s]
2018/10/09 13:11:20 [INFO] [readyfireaim.org] The server validated our request
2018/10/09 13:11:21 [INFO] dreamhost: record_removed
2018/10/09 13:11:21 [INFO] [readyfireaim.org, *.readyfireaim.org] acme: Validations succeeded; requesting certificates
2018/10/09 13:11:22 [INFO] [readyfireaim.org] Server responded with a certificate.
```

**Comment by ldez** [comment]
Could you wait a bit and try again because  `acme: Authorization already valid; skipping challenge` means that the challenge is not executed because of a cache in LE.

Also you have to remove the `.lego` directory (`rm -rf .lego`).

**Comment by cjs** [comment]
Added some live tests. The first run is against a zone I don't own, the second one is one where I have rights:

```console
❯ env DREAMHOST_TEST_DOMAIN="foo.example.com" go test --count=1 github.com/xenolf/lego/providers/dns/dreamhost
2018/10/12 15:13:15 [INFO] dreamhost: record_added
2018/10/12 15:13:15 [INFO] dreamhost: record_removed
--- FAIL: TestLiveDreamHostPresentAndCleanUp (4.85s)
    require.go:794:
                Error Trace:    dreamhost_test.go:147
                Error:          Received unexpected error:
                                dreamhost: dreamhost: add TXT record failed: no_such_zone
                Test:           TestLiveDreamHostPresentAndCleanUp
FAIL
FAIL    github.com/xenolf/lego/providers/dns/dreamhost  4.909s
```

```console
$env DREAMHOST_TEST_DOMAIN="fake.REDACTED_BUT_REAL.com" go test --count=1 github.com/xenolf/lego/providers/dns/dreamhost
ok      github.com/xenolf/lego/providers/dns/dreamhost  6.883s
```

I have also reperformed the test run:

```console
$ env DREAMHOST_PROPAGATION_TIMEOUT=14400 DREAMHOST_POLLING_INTERVAL=60 ./lego -m cjs@readyfireaim.org -a -x http-01 -x tls-alpn-01 --dns dreamhost -s https://acme-staging-v02.api.letsencrypt.org/directory --dns-resolvers 8.8.8.8 -d "*.test.readyfireaim.org" -d test.readyfireaim.org run
2018/10/12 16:12:31 No key found for account cjs@readyfireaim.org. Generating a curve P384 EC key.
2018/10/12 16:12:31 Saved key to /Users/cjs/go/src/github.com/xenolf/lego/.lego/accounts/acme-staging-v02.api.letsencrypt.org/cjs@readyfireaim.org/keys/cjs@readyfireaim.org.key
2018/10/12 16:12:34 [INFO] acme: Registering account for cjs@readyfireaim.org
2018/10/12 16:12:35 !!!! HEADS UP !!!!
2018/10/12 16:12:35
                Your account credentials have been saved in your Let's Encrypt
                configuration directory at "/Users/cjs/go/src/github.com/xenolf/lego/.lego/accounts/acme-staging-v02.api.letsencrypt.org/cjs@readyfireaim.org".
                You should make a secure backup of this folder now. This
                configuration directory will also contain certificates and
                private keys obtained from Let's Encrypt so making regular
                backups of this folder is ideal.
2018/10/12 16:12:35 [INFO] [*.test.readyfireaim.org, test.readyfireaim.org] acme: Obtaining bundled SAN certificate
2018/10/12 16:12:35 [INFO] [*.test.readyfireaim.org] AuthURL: https://acme-staging-v02.api.letsencrypt.org/acme/authz/T0IH-eno5AkYpkJNKvcUH8MV3ld6a3YzdiEG_1xp5JY
2018/10/12 16:12:35 [INFO] [test.readyfireaim.org] AuthURL: https://acme-staging-v02.api.letsencrypt.org/acme/authz/RHjIzW6UIu9tp2Q14K7KRzuZsDZ9C4DKaI7rMYKn4Ts
2018/10/12 16:12:35 [INFO] [test.readyfireaim.org] acme: Preparing to solve DNS-01
2018/10/12 16:12:37 [INFO] dreamhost: record_added
2018/10/12 16:12:37 [INFO] [test.readyfireaim.org] acme: Preparing to solve DNS-01
2018/10/12 16:12:37 [INFO] dreamhost: record_added
2018/10/12 16:12:37 [INFO] [test.readyfireaim.org] acme: Trying to solve DNS-01
2018/10/12 16:12:37 [INFO] [test.readyfireaim.org] Checking DNS record propagation using [8.8.8.8:53]
2018/10/12 16:12:37 [INFO] Wait [timeout: 4h0m0s, interval: 1m0s]
2018/10/12 17:13:05 [INFO] [test.readyfireaim.org] The server validated our request
2018/10/12 17:13:05 [INFO] [test.readyfireaim.org] acme: Trying to solve DNS-01
2018/10/12 17:13:05 [INFO] [test.readyfireaim.org] Checking DNS record propagation using [8.8.8.8:53]
2018/10/12 17:13:05 [INFO] Wait [timeout: 4h0m0s, interval: 1m0s]
2018/10/12 17:13:11 [INFO] [test.readyfireaim.org] The server validated our request
2018/10/12 17:13:11 [INFO] dreamhost: record_removed
2018/10/12 17:13:12 [INFO] dreamhost: record_removed
2018/10/12 17:13:12 [INFO] [*.test.readyfireaim.org, test.readyfireaim.org] acme: Validations succeeded; requesting certificates
2018/10/12 17:13:13 [INFO] [*.test.readyfireaim.org] Server responded with a certificate.
```

Finally, I'd be willing answer questions for users of this DNS provider since I have the ability to test against my account.

**Pull request merged**